python/tests/server/test_worker.py

```diff
@@ -475,57 +510,64 @@ def setup(self):
         try:
             self.setup_generator = self.worker.setup()
             self.setup_events = []
+            # read one event so that we raise InvalidStateException
+            event = self.await_(anext(self.setup_generator))
+            self.setup_events.append(event)
```

Comment:
If you apply my suggestions from `worker.py` and make the signature of `setup` and `predict` a non-async function returning an `AsyncIterator` directly, this hoop-jumping isn't necessary.

Reply:
they'll need to be async functions eventually, I feel like making sync functions that are actually an async generator needlessly confuses the actual code for the sake of the tests

Reply:
It's not just for the sake of the tests. There is a meaningful distinction between a function which does some stuff synchronously and then returns an async iterator, and an async generator function.

The concept of a non-async function returning a `Task` is commonplace in C#, for example. Seems fine. If they eventually need to become async functions, they can eventually become async functions. But let's cross that bridge when we need to and not before.

Reply:
one of the goals for this PR is to make #1405 smaller by getting the simple changes out of the way. predict necessarily has to be async to use the semaphore. making it async requires relatively straightforward changes to the tests. it's easier to make those changes here

as far as I can tell, doing it like this is still correct, so I don't think postponing hoop-jumping that will be necessary anyway benefits us

Reply:
I guess I would accept

```python
async def predict(self, ...): ...

def predict_with_check(self, ...):
        self._assert_state(WorkerState.READY)
        return self.predict(...)
```

Reply:
I still don't think `predict` needs to be an async method. It can do whatever state checks it needs to and then return an async iterator. If there's an inner `async def _predict(...):` and that's where the semaphore acquire goes then that's fine.

Making it `async` and then having to do backflips just to write the tests seems silly.

I think it's perfectly reasonable to make considered choices about whether any particular method or function actually needs to be async or not. Keep in mind that as long as nothing blocks it's totally fine to call non-async functions from within an async context, while the reverse is of course not true.

Reply:
Thank you for detailed discussion on this design choice. I'm happy with this PR as it is, but for the next PR this might be different. To me, having yet another underscore method with the same name for a semaphore _is_ doing backflips in the actual runtime code for the sake of marginally simpler tests, and that seems silly. In practice we're always doing `async for event in worker.predict(...)` and this would add more unnecessary indirection to a codebase that already has a great deal of indirection. 

Python and C# are different languages with different conventions. This doesn't seem pythonic to me, though I will concede that `asyncio.create_task` exists. I spent a little while looking for other examples of this pattern. Claude thought one example might be `httpx.AsyncClient.stream`, but that's actually an async function that starts with a non-async assertion that the protocol is valid.

Again, I'm happy with this PR as is, this is only for the future. I feel like this would be cleaner in the long run, but it's not that important in the grand scheme of things and if you feel really strongly about this I'm ultimately willing to stick to your proposal